Rework light client

jsdw opened the pull request:
The main changes:

- Move almost all light client code into `subxt_lightclient`; only a little glue in Subxt to integrate it.
- `LightClient` struct to create light client instance given relay chain spec, and to create parachain connections on top of this.
- `LightClientRpc` returned from each of the above which is the means to interact with the chain.
- Subxt implements `RpcClientT` for this, so that it can be used to construct an `OnlineClient`.
- Function provided for low level control but not exported yet (maybe best to hide Smoldot away unless somebody has a good reason to need to configure it directly.
- Each light client RPC connection has an associated `BackgroundTask`. This isn't exposed yet, but can eventually be in order that users can drive it and exert backpressure if need be (I'm unsure what effect that would have on Smoldot though!). 

Other changes along the way:
- light client background task is now per-chain-connection so no need to combine and then separate messages from different chains, and in the future we can opt to drive each one manually if we want.
- Handle error messages on subscriptions, too.
- Generally split errors up so that we get quite specific erros back when constructing, calling or getting responses back from the light client stuff.

Things not done:
- Exposing something like `OnlineClient::from_chain_spec()`; need a relay + parachain version etc, prob need a builder or something to make it nice. Instead I made it easier to instantiate an OnlineClient from an RPC client (`impl Into`) so it's actually pretty easy anyway now.
- I didn't add a nice `from_url` to the `ChainSpecConfig`. I staretd doing this but it involved pulling in extra deps and feature flags etc, and the code was just easier to keep in `subxt`. So I exposed a utility function to fetch chain spec from url instead. I'm ok with this because it's actually only really a dev use case anyway; we prob shouldn't encourage it too much!

jsdw committed:
WIP second pass over light client code for simpler API

jsdw committed:
First pass new light client

jsdw committed:
pub(crate) LightClientRpc::new_raw(), and fmt

jsdw committed:
Update examples and add back a way to configure boot nodes and fetch chainspec from a URL

jsdw committed:
Fix light client examples

jsdw committed:
remove unused deps and tidy lightclient feature flags

jsdw committed:
fix wasm error

jsdw committed:
LightClientRpc can be cloned

jsdw committed:
update light client tests

jsdw commented on lightclient/src/background.rs:
```diff
@@ -1,43 +1,47 @@
+>;
+
+/// Type of subscription IDs we can get back.
+pub type SubscriptionId = String;
```
It was previously an unsigned int, but even if Smoldot _currently_ does that, I didn't think we should rely on it since it probably also expects us to treat the valeus as opaque.

jsdw commented on lightclient/src/background.rs:
```diff
@@ -46,29 +50,160 @@ pub enum FromSubxt {
+    /// Run the background task, which:
+    /// - Forwards messages/subscription requests to Smoldot from the front end.
+    /// - Forwards responses back from Smoldot to the front end.
+    pub async fn run(self) {
```
A simplified version of what was called `start_task` before. No longer takes in any args, so the API for `BackgroundTask` is super simple; just `.run().await` it to start it running (and in the future we can return something users can poll which will just wrap this).

jsdw commented on lightclient/src/background.rs:
```diff
@@ -46,29 +50,160 @@ pub enum FromSubxt {
+                let from_back_fut = channels.from_back.next().fuse();
+            }
+
+            futures::select! {
```
I moved to this I think to avoid allocating somewhere or annoying lifetime things; can't remember exactly now. It just takes two channels as inputs though and so I think should have no cancellation concerns (this is the usual footgun with `select!`

tadeohepperle replied:
Looks good, btw you asked me exactly this question "What is the usual usual footgun with select!" when interviewing me as a job applicant about a year ago.

jsdw replied:
I need to make a note somewhere because if I move to doing a code review style task then `future::select!` is def a candidate to put somewhere :D

niklasad1 replied:
you are good, streams are cancel-safe.

I like tokio::select better because it doesn't require pinning the futures

jsdw replied:
I normally prefer the function call version to avoid some footguns but iirc I had an issue that made it hard to use it nicely; I'll have to try again at some point!

jsdw commented on lightclient/src/background.rs:
```diff
@@ -100,87 +227,67 @@ struct PendingSubscription {
 struct ActiveSubscription {
     /// Channel to send the subscription notifications back to frontend.
-    sender: mpsc::UnboundedSender<Box<RawValue>>,
+    notification_sender: mpsc::UnboundedSender<Result<Box<RawValue>, JsonRpcError>>,
```
I tweaked subscriptions a bit to:

a) return any subscription notification errors that come back (previously these were being ignored I think) and
b) send back the ID and repsonse channel rather than one channel for ID and one for responses, just to simplify usage a little.

jsdw commented on lightclient/src/chain_config.rs:
```diff
@@ -0,0 +1,71 @@
+}
+
+/// Configuration to connect to a chain.
+pub struct ChainConfig<'a> {
```
This can grow to have more options like about max json RPC subs etc if we want to expose them. A user can then either build this config and configure it, or pass a chain spec `&str` directly, which will convert into this easily.

jsdw commented on lightclient/src/rpc.rs:
```diff
@@ -0,0 +1,132 @@
+        /// Result.
+        result: Box<RawValue>,
+    },
+    NotificationError {
+        /// RPC method that generated the notification.
+        method: String,
+        /// Subscription ID.
+        subscription_id: String,
+        /// Result.
+        error: Box<RawValue>,
+    },
```
This file is copied from the previous impl in `background.rs`, but I added this `NotificationError` to handle the case when we get errors back in subscriptions. Worth double checking that this is right :)

jsdw commented on subxt/src/utils/fetch_chain_spec.rs:
```diff
@@ -0,0 +1,114 @@
+
+cfg_jsonrpsee! {
+    /// Fetch a chain spec from an RPC node at the given URL.
+    pub async fn fetch_chainspec_from_rpc_node(url: impl AsRef<str>) -> Result<Box<RawValue>, FetchChainspecError> {
```
I tried putting a fn like this (`from_url`) on `ChainConfig` but it required pulling lots of deps in etc, and since it's a dev use case mostly I was sortof ok with just exposing this util function instead in Subxt, since it's easier to do it here. Not the cleanest though!

jsdw committed:
Other small fixes

jsdw committed:
exclude mod unless jsonrpsee

jsdw committed:
Fix wasm-lightclient-tests

jsdw committed:
Merge branch 'master' into jsdw-lightclient-2

lexnv commented on subxt/examples/light_client_parachains.rs:
```diff
@@ -16,85 +12,34 @@ const ASSET_HUB_SPEC: &str =
-    // Connecting to a parachain is a multi step process.
+    // Instantiate a light client with the Polkadot relay chain,
+    // and connect it to Asset Hub, too.
+    let (lightclient, polkadot_rpc) = LightClient::relay_chain(POLKADOT_SPEC)?;
```
Looks amazing!

lexnv reviewed (approved):
LGTM! I like the idea of having a background task per chain ID, this should isolate things quite nicely 👍

jsdw committed:
add back docsrs bit and web+native feature flag compile error

jsdw committed:
update book and light client example names

jsdw committed:
Merge branch 'master' into jsdw-lightclient-2

jsdw commented:
I added back a compile error and docsrs thing that went missing, and updated the book, swapping the examples around so now connecting to parachains is the basic example, and connecting to a local node is more advanced :)

jsdw committed:
fix docs

tadeohepperle commented on lightclient/src/background.rs:
```diff
@@ -100,87 +227,67 @@ struct PendingSubscription {
     /// Send the method response ID back to the user.
```
```suggestion
    /// the method response ID back to the user.
```

tadeohepperle commented on lightclient/src/background.rs:
```diff
@@ -253,129 +357,133 @@ impl<TPlatform: PlatformRef, TChain> BackgroundTask<TPlatform, TChain> {
+                    // Send the method response and a channel to receive notifications back.
+                    if pending_subscription
+                        .response_sender
+                        .send(Ok((sub_id.clone(), sub_rx)))
```
Could we avoid allocation here by using `Arc<str>` for subscription ids? Probably irrelevant though, feel free to ignore.

jsdw replied:
I'd guess it's not worth while because we (technically, but not really in reality) need to allocate a `String` anyway for the deserializing bit, and then the `Arc` is another alloc, and we onlyclone it once anyways :)

tadeohepperle commented on lightclient/src/lib.rs:
```diff
@@ -2,52 +2,259 @@
+    /// https://docs.rs/wasm-bindgen-futures/latest/wasm_bindgen_futures/fn.future_to_promise.html.
+    pub fn parachain<'a>(
+        &self,
+        chain_config: impl Into<ChainConfig<'a>>,
```
This interface with the `relay_chain` and `parachain` functions looks really clean.

tadeohepperle reviewed (approved):
I think the interface is well designed, good job!

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/parity-tech-update-for-march/7226/1
